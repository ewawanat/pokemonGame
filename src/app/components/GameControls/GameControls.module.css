.controlsContainer {
    text-align: center;
    margin-top: 20px;
}

.controlButton {
    padding: 10px 20px;
    background-color: #007bff;
    color: white;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s ease;
    font-size: 1rem;
}

.controlButton:hover {
    background-color: #0056b3;
}

.controlButton:active {
    background-color: #003f7f;
}

.score {
    font-size: 1.2rem;
    font-weight: bold;
    color: #333;
    margin-top: 10px;
}

.results {
    margin-top: 20px;
    text-align: center;
}

.finalScore {
    font-size: 1.5rem;
    color: #007bff;
    margin-bottom: 10px;
}