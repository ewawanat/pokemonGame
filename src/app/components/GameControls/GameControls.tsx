import React from 'react';
import styles from './GameControls.module.css';

type GameControlsProps = {
    isGameStarted: boolean;
    isGameFinished: boolean;
    score: number;
    totalQuestions: number;
    correctAnswers: number;
    startGame: () => void;
    finishGame: () => void;
};

const GameControls: React.FC<GameControlsProps> = ({
    isGameStarted,
    isGameFinished,
    score,
    totalQuestions,
    correctAnswers,
    startGame,
    finishGame,
}) => {
    const calculatePercentage = () => {
        return totalQuestions > 0 ? ((correctAnswers / totalQuestions) * 100).toFixed(2) : '0';
    };

    return (
        <div className={styles.controlsContainer}>
            <button
                onClick={!isGameStarted || isGameFinished ? startGame : finishGame}
                className={styles.controlButton}
            >
                {!isGameStarted || isGameFinished ? 'Start Game' : 'Finish Game'}
            </button>
            <p className={styles.score}>Score: {score}</p>
            {isGameFinished && (
                <div className={styles.results}>
                    <h2 className={styles.finalScore}>Final Score: {score}</h2>
                    <p>Correct Answers: {correctAnswers} / {totalQuestions}</p>
                    <p>Percentage: {calculatePercentage()}%</p>
                </div>
            )}
        </div>
    );
};

export default GameControls;
