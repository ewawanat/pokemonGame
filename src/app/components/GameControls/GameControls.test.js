import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';
import GameControls from './GameControls';

// Extend Jest to include axe accessibility checks
expect.extend(toHaveNoViolations);

describe('GameControls Component', () => {
    const mockStartGame = jest.fn();
    const mockFinishGame = jest.fn();

    const defaultProps = {
        isGameStarted: false,
        isGameFinished: false,
        score: 0,
        totalQuestions: 10,
        correctAnswers: 0,
        startGame: mockStartGame,
        finishGame: mockFinishGame,
        questionsAnswered: 0,
    };

    const renderComponent = (props = {}) => {
        return render(<GameControls {...defaultProps} {...props} />);
    };

    it('renders the "Start Game" button when the game is not started or finished', () => {
        renderComponent();

        const startButton = screen.getByRole('button', { name: /start game/i });
        expect(startButton).toBeInTheDocument();
        expect(startButton).toHaveTextContent('Start Game');
    });

    it('renders the "Finish Game" button when the game is started and not finished', () => {
        renderComponent({ isGameStarted: true });

        const finishButton = screen.getByRole('button', { name: /finish game/i });
        expect(finishButton).toBeInTheDocument();
        expect(finishButton).toHaveTextContent('Finish Game');
    });

    it('calls startGame when the "Start Game" button is clicked', async () => {
        renderComponent();

        const startButton = screen.getByRole('button', { name: /start game/i });
        await userEvent.click(startButton);

        expect(mockStartGame).toHaveBeenCalledTimes(1);
    });

    it('calls finishGame when the "Finish Game" button is clicked', async () => {
        renderComponent({ isGameStarted: true });

        const finishButton = screen.getByRole('button', { name: /finish game/i });
        await userEvent.click(finishButton);

        expect(mockFinishGame).toHaveBeenCalledTimes(1);
    });

    it('displays the score correctly', () => {
        renderComponent({ score: 50 });

        const scoreText = screen.getByText(/score: 50/i);
        expect(scoreText).toBeInTheDocument();
    });

    it('displays the final results when the game is finished', () => {
        renderComponent({
            isGameFinished: true,
            score: 80,
            correctAnswers: 8,
            totalQuestions: 10,
        });

        const finalScore = screen.getByText(/final score: 80/i);
        const correctAnswersText = screen.getByText(/correct answers: 8 \/ 10/i);
        const percentageText = screen.getByText(/percentage: 80.00%/i);

        expect(finalScore).toBeInTheDocument();
        expect(correctAnswersText).toBeInTheDocument();
        expect(percentageText).toBeInTheDocument();
    });

    it('does not display final results when the game is not finished', () => {
        renderComponent();

        const finalScore = screen.queryByText(/final score:/i);
        expect(finalScore).not.toBeInTheDocument();
    });

    it('matches the snapshot when the game is not started', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches the snapshot when the game is finished', () => {
        const { asFragment } = renderComponent({
            isGameFinished: true,
            score: 80,
            correctAnswers: 8,
            totalQuestions: 10,
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('has no accessibility violations', async () => {
        const { container } = renderComponent();
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });

    it('has no accessibility violations when the game is finished', async () => {
        const { container } = renderComponent({
            isGameFinished: true,
            score: 80,
            correctAnswers: 8,
            totalQuestions: 10,
        });
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
