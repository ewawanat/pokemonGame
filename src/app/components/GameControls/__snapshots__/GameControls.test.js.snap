// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControls Component matches the snapshot when the game is finished 1`] = `
<DocumentFragment>
  <div
    class="controlsContainer"
  >
    <button
      class="controlButton"
    >
      Start Game
    </button>
    <p
      class="score"
    >
      Score: 80
    </p>
    <div
      class="results"
    >
      <h2
        class="finalScore"
      >
        Final Score: 80
      </h2>
      <p>
        Correct Answers: 8 / 10
      </p>
      <p>
        Percentage: 80.00%
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GameControls Component matches the snapshot when the game is not started 1`] = `
<DocumentFragment>
  <div
    class="controlsContainer"
  >
    <button
      class="controlButton"
    >
      Start Game
    </button>
    <p
      class="score"
    >
      Score: 0
    </p>
  </div>
</DocumentFragment>
`;
